Update MonaLisa code in WMCore. fixes #6301

juztas opened the pull request:
This also brings ProcInfo.py and Logger.py which is used in CRAB3. Later it will be removed from CRAB3 and will use from WMCore.

juztas committed:
Update MonaLisa code in WMCore. fixes #6301
This also brings ProcInfo.py and Logger.py which is used in CRAB3. Later it will be removed from CRAB3 and will use from WMCore.

ticoann commented:
@juztas, Hi Justas, I have no idea how this works. Since this is used by CRAB3 already, I assume this is tested. Is there a unittest for in the CRAB3 code as well?

juztas commented:
No, there is no unit test. Let me first get answer from dashboard team about endhosts and I have some tests files ready. Will need to rewrite them to unit tests

juztas commented:
I got reply from Julia and DashboardAPI can also be moved to WMCore. So leave it still open, I will try to update it this week

ticoann commented:
@juztas, Hi Justas, what is the status of this patch. Can this be merged?

juztas commented:
wait, will also modify dashboardAPI which has few more functions in crab.

ericvaandering commented:
test this please

ticoann commented:
@juztas, Thank you very much Justas, actually I don't know much about this to really review the code.
One thing is if you could move the test parts to under unittest area

https://github.com/dmwm/WMCore/tree/master/test/python/WMCore_t/Services_t
The structue is the unitest file name uses the same name as original file name _t.
Sorry for the late comment

ericvaandering commented:
I wondered if we satisfy all the dependencies. It looks like this thing needs “future” (where we get futurize) but we may just have that in the devtools RPM and not in the actual WMCore.

ticoann commented:
Ah, that is good point, Eric. For agent rpm, future is already added. But I don't think it is added in crabserver or client. Whichever uses this library, need to have dependency set in its rpm spec file.

ericvaandering commented:
The tests are in a file we get directly from MonaLisa and not really written as a unit test. So rather than moving them to our testing, what I think is an easy way to do this is to write a test under our structure that simply does (pseudoCode):

```
from Logger import main as loggerMain

def testLogger():
    loggerMain()

juztas commented:
Sorry, missed all discussion, what is left to move forward ? I would like to move this reporting to dashboard about stageout issues in next release.

CS also has future added in their apmon. Maybe @mmascher can double check if it is included.

ericvaandering commented:
Well, I'd like you to add a unit test in WMCore as I suggested 2 comments above. Then we can merge this into WMCore.

juztas committed:
Unittest for DashboardAPI

juztas commented:
Ok, Added few fixes to new unit test. I still have few questions:
a) Unittests are running with python2.7? This is what I see from the logs, so I could use unittest features which are only available in python2.7
b) I had question about the path which I should look for a file, but I added this `os.path.dirname(os.path.abspath(__file__))` and lets see how jenkins will react.
c) WMCore enforces to use 
```
Tue Feb 23 13:53:09 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Tue Feb 23 13:53:09 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
```
Is there running Monalisa repository on test machine which collects all parameters?

juztas commented:
Thanks @ericvaandering changes done, I added one additional print as I am not sure who rewrites `APMONCONF`. Let see...

ericvaandering commented on test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py:
```diff
@@ -100,14 +100,14 @@ def testLHEProductionWorkload(self):
                     last_lumi = first_lumi
                     last_lumi += math.ceil(((last_event - first_event + 1) %
                                 SliceSize) / splitArgs['SubSliceSize'])
-                    last_lumi += (lumis_per_job * (unit['Jobs'] - 1)) - 1
```
What in the world do these changes have to do with the dashboard?

ericvaandering commented:
I don't see how the MonteCarlo bit can be related whatsoever. This should not be merged. It may be that you are getting blamed for something else. What I'd like to see you do is get rid of the changes to MonteCarlo_t and rebase this test onto the latest master and then we can see.

juztas commented:
Updated and failures does not seem to be related to this pull request.

juztas commented:
@ericvaandering @ticoann comments?

ericvaandering commented:
test this please

ericvaandering commented:
The WMCore_t.WorkQueue_t.Policy_t.Start_t.MonteCarlo_t errors are real. Please take a look

juztas commented:
Mh... I am still quite not sure how my pull request can affect MonteCarlo.
Just maybe for my understanding: I tried to grep through the git history and not found any failure related MonteCarlo_t failure. What was the last time it ran successfully? How I could enforce to run it on jenkins? I don't have WMA env to test it locally.

ericvaandering commented:
https://cmssdt.cern.ch/jenkins/job/DMWM-WMCore-PR-test/1823/testReport/junit/WMCore_t.WorkQueue_t.Policy_t.Start_t.MonteCarlo_t/MonteCarloTestCase/testShiftedStartSplitting/ shows you the debug output of what's going on with the test. It runs successfully every night.

https://github.com/dmwm/WMCore/wiki/Setup-wmcore-unittest explains how to setup your own test environment for WMCore. Its quite easy.

juztas commented:
test this please

juztas commented:
Finally I have my vm ready, but tests are succeeding :/ Maybe I missed something. I just want to force to get the log file again from jenkins

juztas commented:
```
$ python test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py
Creating ApMon with static configuration
Wed Mar 30 17:18:54 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:18:54 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:18:54 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:18:55 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
..Creating ApMon with static configuration
Wed Mar 30 17:18:55 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:18:55 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:18:55 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:18:56 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
.Creating ApMon with static configuration
Wed Mar 30 17:18:56 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:18:56 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:18:56 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:18:57 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
Creating ApMon with static configuration
Wed Mar 30 17:18:57 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:18:57 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:18:58 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:18:59 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
.Creating ApMon with static configuration
Wed Mar 30 17:18:59 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:18:59 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:18:59 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:19:00 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
...Creating ApMon with static configuration
Wed Mar 30 17:19:00 2016: ApMon[ERROR]: Cannot open 127.0.0.1:8884
Wed Mar 30 17:19:00 2016: ApMon[ERROR]: IOError: [Errno 2] No such file or directory: '127.0.0.1:8884'
Wed Mar 30 17:19:00 2016: ApMon[WARNING]: Not sending parameters since no destination is defined.
Wed Mar 30 17:19:01 2016: ApMon[DEBUG]: Closing UDP socket on ApMon object destroy.
.
----------------------------------------------------------------------
Ran 8 tests in 7.411s

OK
[jbalcas@gwmsmon-development WMCore]$ git status
# On branch new_apmon
```
I also tried to run all unittests as jenkin is doing, but it got stuck on my VM. As I am not an expert of WorkQueue, maybe someone could help me to figure out these faster :) ?

ericvaandering commented:
How about an intermediate test as mentioned in the testing Twiki: 
```
python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py
```
That should do exactly what Jenkins is doing but running only the tests you want (and not any possible interactions between them)

juztas commented:
```
$ python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py
running test
Using the tests below: test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py
#### WE ARE DELETING YOUR DATABASE. 3 SECONDS TO CANCEL ####
#### buildbotmode is true
We are going to trash databases after every test
### We are in buildbot mode ###
----------------------------------------------------------------------
Ran 8 tests in 0.001s

OK
path lists is ['test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py']
Out of 8 cases, we will run 8
#1 Basic Production Workload ... ok
#2 Doesn't support continuous splitting ... ok
#3 _testExtremeSplits_ ... ok
#4 Specs with no work ... ok
#5 _testLHEProductionWorkload_ ... ok
#6 Multi merge production workload ... ok
#7 Multi Task Processing Workflow ... ok
#8 _testShiftedStartSplitting_ ... ok

----------------------------------------------------------------------
XML: /data/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 8 tests in 7.540s

OK
Testing complete, there are now 1 threads
```

ticoann commented:
@juztas, @ericvaandering, Is this ready to be merged?

ericvaandering commented:
IMO, no. It's causing a test in Jenkins to regularly fail. I want to try to understand why.

ericvaandering commented:
Ok, what if you do the test with the correct slice. Is that what freezes?

juztas commented:
rrr. after debugging found it. It is float vs int division and what math.ceil is returning. Now why jenkins gets everything as integer and locally as float still not clear for me. I will provide a fix for these two and lets see jenkins output ;)

juztas committed:
Int vs float division with math.ceil different results

juztas commented:
Just for reference:
```
>>> import math
>>> last_event = 542674
>>> last_lumi = 46181
>>> first_event = float(542616)
>>> SliceSize = 47
>>> SubSlice = 13
>>> lumis_per_job = 4
>>> jobsunits = 2
>>> last_lumi += math.ceil(((last_event - first_event + 1) % SliceSize) / SubSlice)
>>> last_lumi += (lumis_per_job * (jobsunits - 1)) - 1
>>> print last_lumi # returns 46185
46185.0
>>> import math
>>> last_event = 542674
>>> last_lumi = 46181
>>> first_event = 542616
>>> SliceSize = 47
>>> SubSlice = 13
>>> lumis_per_job = 4
>>> jobsunits = 2
>>> last_lumi += math.ceil(((last_event - first_event + 1) % SliceSize) / SubSlice)
>>> last_lumi += (lumis_per_job * (jobsunits - 1)) - 1
>>> print last_lumi # returns 46184
46184.0
```

ericvaandering commented on test/python/WMCore_t/WorkQueue_t/Policy_t/Start_t/MonteCarlo_t.py:
```diff
@@ -88,7 +88,7 @@ def testLHEProductionWorkload(self):
             self.assertEqual(math.ceil(float(totalevents) / (SliceSize * splitArgs['MaxJobsPerElement'])),
                              len(units))
-            first_event = 1
+            first_event = 1.0
```
That is really weird. In any case, it seems this is a good time to 

```
from __future__ import division
``` 

for this test rather than mucking around with promoting things to get python 2 to do the right thing. Can you try that instead?

juztas replied:
sure, let me try first locally

ericvaandering commented:
If you find that different things now fail, you should be able to replace / with // to restore python2 integer division

juztas commented:
@ericvaandering is there anything more with this? I want to remake this https://github.com/dmwm/WMCore/pull/6537 pull request and have this in Dashboard implemented. This should give nice view about stageout success rate ;)

ericvaandering commented:
I'm OK with it now. Thanks for your patience.

juztas commented:
Thanks to you! I learned more stuff doing it :)

ticoann commented on src/python/WMCore/Services/Dashboard/Logger.py:
```diff
@@ -0,0 +1,78 @@
+"""
+from __future__ import print_function
+from __future__ import division
+from builtins import object
```
Hi Justas, builtins seems not in python 2.7 standard library. and this is causing the problem in production agent. 
Is this import needed?

ticoann replied:
To be more clear, the problem occurs since the code is running in worker node. If the code is run agent there is no problem, since agent rpm has dependency on future package. But it doesn't seem that package is ported to worker node.  Anyway, I will create a new issue for that.
